common/port_manager_test.cc

```diff
@@ -92,109 +93,137 @@ TEST_F(PortManagerTest, ReservePortAllLocalPortsTaken) {
 
 TEST_F(PortManagerTest, ReleasePort) {
-  process_factory_.SetProcessOutput(kLocalNetstat, "", "", 0);
-  process_factory_.SetProcessOutput(kRemoteNetstat, "", "", 0);
+  process_factory_.SetProcessOutput(kWindowsNetstat, "", "", 0);
+  process_factory_.SetProcessOutput(kLinuxNetstat, "", "", 0);
 
-  absl::StatusOr<int> port = port_manager_.ReservePort(kTimeoutSec);
+  absl::StatusOr<int> port =
+      port_manager_.ReservePort(kTimeoutSec, ArchType::kLinux_x86_64);
   EXPECT_EQ(*port, kFirstPort);
   EXPECT_OK(port_manager_.ReleasePort(*port));
-  port = port_manager_.ReservePort(kTimeoutSec);
+  port = port_manager_.ReservePort(kTimeoutSec, ArchType::kLinux_x86_64);
   EXPECT_EQ(*port, kFirstPort);
 }
 
 TEST_F(PortManagerTest, ReleasePortOnDestruction) {
-  process_factory_.SetProcessOutput(kLocalNetstat, "", "", 0);
-  process_factory_.SetProcessOutput(kRemoteNetstat, "", "", 0);
+  process_factory_.SetProcessOutput(kWindowsNetstat, "", "", 0);
+  process_factory_.SetProcessOutput(kLinuxNetstat, "", "", 0);
 
   auto port_manager2 = std::make_unique<PortManager>(
       kGuid, kFirstPort, kLastPort, &process_factory_, &remote_util_);
-  EXPECT_EQ(*port_manager2->ReservePort(kTimeoutSec), kFirstPort + 0);
-  EXPECT_EQ(*port_manager_.ReservePort(kTimeoutSec), kFirstPort + 1);
+  EXPECT_EQ(*port_manager2->ReservePort(kTimeoutSec, ArchType::kLinux_x86_64),
+            kFirstPort + 0);
+  EXPECT_EQ(*port_manager_.ReservePort(kTimeoutSec, ArchType::kLinux_x86_64),
+            kFirstPort + 1);
   port_manager2.reset();
-  EXPECT_EQ(*port_manager_.ReservePort(kTimeoutSec), kFirstPort + 0);
+  EXPECT_EQ(*port_manager_.ReservePort(kTimeoutSec, ArchType::kLinux_x86_64),
+            kFirstPort + 0);
 }
 
-TEST_F(PortManagerTest, FindAvailableLocalPortsSuccess) {
-  // First port is taken
+TEST_F(PortManagerTest, FindAvailableLocalPortsSuccessWindows) {
+  // First port is in use.
   std::string local_netstat_out =
-      absl::StrFormat(kLocalNetstatOutFmt, kFirstPort);
-  process_factory_.SetProcessOutput(kLocalNetstat, local_netstat_out, "", 0);
+      absl::StrFormat(kWindowsNetstatOutFmt, kFirstPort);
+  process_factory_.SetProcessOutput(kWindowsNetstat, local_netstat_out, "", 0);
 
   absl::StatusOr<std::unordered_set<int>> ports =
-      PortManager::FindAvailableLocalPorts(kFirstPort, kLastPort, "127.0.0.1",
-                                           &process_factory_);
+      PortManager::FindAvailableLocalPorts(
+          kFirstPort, kLastPort, ArchType::kWindows_x86_64, &process_factory_);
+  ASSERT_OK(ports);
+  EXPECT_EQ(ports->size(), kNumPorts - 1);
+  for (int port = kFirstPort + 1; port <= kLastPort; ++port) {
+    EXPECT_TRUE(ports->find(port) != ports->end());
+  }
+}
+
+TEST_F(PortManagerTest, FindAvailableLocalPortsSuccessLinux) {
+  // First port is in use.
+  std::string local_netstat_out =
+      absl::StrFormat(kLinuxNetstatOutFmt, kFirstPort);
+  process_factory_.SetProcessOutput(kLinuxNetstat, local_netstat_out, "", 0);
```

Comment:
I'm not sure if I understand what's happening here. Are you using `netstat` to determine which ports are in use?
If so, what about race conditions?

Reply:
This dates back to the Stadia days, where we assumed that a certain range of ports was reserved for Stadia, so nobody outside of Stadia would use those ports. It was also assumed that only one Windows device controls a gamelet. To prevent that multiple instances of cdc_rsync (on the same Windows device) reserve the same ports when they run concurrently, PortManager uses inter-process shared memory to guarantee no races.

Nowadays, those assumptions don't necessarily hold anymore, so I'm investigating alternatives to calling netstat. For instance, the server can just try to bind a socket. If it succeeds, the port is available. This requires some coordination, though, since we can't run port forwarding before running the server anymore:
1) server binds port, reports to client
2) client sets up SSH port forwarding
3) client connects
4) server accepts connection
It's still unclear to me whether you can bind to a port before setting up port forwarding.